server/src/microservices/processors/metadata-extraction.processor.ts

```diff
@@ -100,6 +114,102 @@ export class MetadataExtractionProcessor {
     }
   }
 
+  private async extractPixelLivePhoto(
+    asset: AssetEntity,
+    directory: DirectoryEntry[],
+    fileCreatedAt: Date | null,
+  ): Promise<AssetEntity | null> {
+    if (asset.livePhotoVideoId) {
+      // Already extracted, don't try again.
+      const [ret] = await this.assetRepository.getByIds([asset.livePhotoVideoId]);
+      this.logger.log(`Already extracted asset ${ret.originalPath}.`);
+      return ret;
+    }
+    let foundMotionPhoto = false;
+    let motionPhotoOffsetFromEnd = 0;
+    let motionPhotoLength = 0;
+    let motionPhotoMime = '';
+
+    // Look for the directory entry with semantic label "MotionPhoto", which is the embedded video.
+    // Then, determine the length from the end of the file to the start of the embedded video.
+    for (let i = 0; i < directory.length; i++) {
+      if (directory[i].Item.Semantic == 'MotionPhoto') {
+        if (foundMotionPhoto) {
+          this.logger.error(`Asset ${asset.originalPath} has more than one motion photo.`);
+          return null;
+        }
+        foundMotionPhoto = true;
+        motionPhotoLength = directory[i].Item.Length ?? 0;
+        motionPhotoMime = directory[i].Item.Mime;
+      }
+      if (foundMotionPhoto) {
+        motionPhotoOffsetFromEnd += directory[i].Item.Length ?? 0;
+        motionPhotoOffsetFromEnd += directory[i].Item.Padding ?? 0;
+      }
+    }
+
+    if (!foundMotionPhoto || motionPhotoLength == 0) {
+      return null;
+    }
+    const file = await fs.promises.open(asset.originalPath).catch((err) => {
+      this.logger.error(`Failed to open asset ${asset.originalPath}: ${err}`);
+      return null;
+    });
+    if (!file) {
+      return null;
+    }
+    let extracted = null;
+    try {
+      // Read in embedded video.
+      const stat = await file.stat().catch((err) => {
+        this.logger.error(`Failed to stat asset ${asset.originalPath}: ${err}`);
+        return null;
+      });
+      if (!stat) {
+        return null;
+      }
+      const motionPhotoOffsetFromStart = stat.size - motionPhotoOffsetFromEnd;
+      extracted = await file
+        .read({
+          buffer: Buffer.alloc(motionPhotoLength),
+          position: motionPhotoOffsetFromStart,
+          length: motionPhotoLength,
+        })
+        .catch((err) => {
+          this.logger.error(`Failed to read embedded video: ${err}`);
+          return null;
+        });
+
+      if (!extracted) {
+        return null;
+      }
+    } finally {
+      await file.close();
+    }
+
+    // Write out extracted video, and add it to the asset repository.
+    const livePhotoPath = asset.originalPath + '.mp4';
```

Comment:
Currently this is writing the extracted video to the same location as the original asset. Since this is generated by Immich, and we have to account for read-only libraries, it's better if this was stored in a separate location. I think the same location as encoded videos makes sense.

Reply:
I've tried to implement this, but the storage template gets applied so the video always gets moved to the library folder. That might not be an issue though, since for read-only libraries I assume that wouldn't happen?

Reply:
Correct, can you use the storage.core.ts file to get the path to encoded videos folder and put it there instead?